kernel/src/schema/mod.rs

```diff
@@ -212,33 +215,104 @@ impl StructField {
                 field: &'a StructField,
             ) -> Option<Cow<'a, StructField>> {
                 let field = self.recurse_into_struct_field(field)?;
-                Some(Cow::Owned(field.with_name(field.physical_name())))
+
+                let metadata = field.logical_to_physical_metadata(self.column_mapping_mode);
+                let name = match self.column_mapping_mode {
+                    ColumnMappingMode::None => field.name().to_owned(),
+                    ColumnMappingMode::Id | ColumnMappingMode::Name => {
+                        // Assert that the physical name is present
+                        debug_assert!(field
+                            .metadata
+                            .get(ColumnMetadataKey::ColumnMappingPhysicalName.as_ref())
+                            .is_some_and(|x| matches!(x, MetadataValue::Number(_))));
+                        field.physical_name().to_owned()
+                    }
+                };
+
+                Some(Cow::Owned(StructField {
+                    name,
+                    data_type: field.data_type.clone(),
+                    nullable: field.nullable,
+                    metadata,
+                }))
             }
         }
         // NOTE: unwrap is safe because the transformer is incapable of returning None
-        MakePhysical
-            .transform_struct_field(self)
-            .unwrap()
-            .into_owned()
+        MakePhysical {
+            column_mapping_mode,
+        }
+        .transform_struct_field(self)
+        .unwrap()
+        .into_owned()
     }
 
     fn has_invariants(&self) -> bool {
         self.metadata
             .contains_key(ColumnMetadataKey::Invariants.as_ref())
     }
+
+    /// Converts a logical schema's StructField metadata into a physical metadata base on the
+    /// provided `column_mapping_mode`.
+    ///
+    /// NOTE: Caller affirms that the schema was already validated by
+    /// [`crate::table_features::validate_schema_column_mapping`], to ensure that annotations are
+    /// always and only present when column mapping mode is enabled.
+    fn logical_to_physical_metadata(
+        &self,
+        column_mapping_mode: ColumnMappingMode,
+    ) -> HashMap<String, MetadataValue> {
+        let mut base_metadata = self.metadata.clone();
+        let physical_name_key = ColumnMetadataKey::ColumnMappingPhysicalName
+            .as_ref()
+            .to_string();
+        let field_id_key = ColumnMetadataKey::ColumnMappingId.as_ref().to_string();
+        let parquet_field_id_key = ColumnMetadataKey::ParquetFieldId.as_ref().to_string();
+
+        let field_id = base_metadata.get(&ColumnMetadataKey::ColumnMappingId.as_ref().to_string());
+        match column_mapping_mode {
+            ColumnMappingMode::Id => {
+                let Some(MetadataValue::Number(fid)) = field_id else {
+                    panic!("StructField with name {} is missing field id in the Id column mapping mode", self.name())
```

Comment:
Note: this is an invariant violation that `validate_schema_column_mapping` should've caught.